californium-core/src/main/java/org/eclipse/californium/core/network/CoapEndpoint.java

```diff
@@ -465,8 +519,28 @@ public void sendEmptyMessage(final Exchange exchange, final EmptyMessage message
 	 *  @throws NullPointerException if the given deliverer is {@code null}
 	 */
 	@Override
-	public void setMessageDeliverer(MessageDeliverer deliverer) {
-		coapstack.setDeliverer(deliverer);
+	public void setMessageDeliverer(final MessageDeliverer deliverer) {
+		if (deliverer == null) {
+			throw new NullPointerException("Message deliverer must not be null");
+		} else {
+			coapstack.setDeliverer(new MessageDeliverer() {
+
+				@Override
+				public void deliverResponse(final Exchange exchange, final Response response) {
+					deliverer.deliverResponse(exchange, response);
+					if (response.isNotification()) {
+						for (NotificationListener listener : notificationListeners) {
```

Comment:
If this is related to Simon's "to be able to get notification from any Californium instance", should these listeners represent each other node in the cluster? And each listener has acommunication adapter to reach the other node (in different container, VM, machine, ...)?

I would have expected something analogous to the ClusterObservationStore that takes over a consistent, ordered reception of notifications. It would basically be a distributed NotificationOrderer that is coordinated over all custer nodes. (See other comment on how it should be quite similar to the current NotificationOrderer.)